source/ImageSet.cpp

```diff
@@ -24,76 +24,76 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include <algorithm>
 #include <cassert>
-#include <iterator>
 
 using namespace std;
 
 namespace {
-	// Determine whether the given path is to an @2x image.
-	bool Is2x(const string &path)
-	{
-		if(path.length() < 7)
-			return false;
-
-		size_t pos = path.length() - 7;
-		return (path[pos] == '@' && path[pos + 1] == '2' && path[pos + 2] == 'x');
-	}
-
-	// Determine whether the given path is to a swizzle mask.
-	bool IsSwizzleMask(const string &path, bool is2x)
-	{
-		if(path.length() < 7 || (is2x && path.length() < 10))
-			return false;
-
-		size_t pos = path.length() - (is2x ? 10 : 7);
-		return (path[pos] == '@' && path[pos + 1] == 's' && path[pos + 2] == 'w');
-	}
-
+	const set<string> SUPPORTED_EXTENSIONS{"png", "jpg", "jpeg"};
 	// Check if the given character is a valid blending mode.
 	bool IsBlend(char c)
 	{
 		return (c == '-' || c == '~' || c == '+' || c == '=');
 	}
 
-	// Determine whether the given path or name is to a sprite for which a
-	// collision mask ought to be generated.
-	bool IsMasked(const string &path)
-	{
-		if(path.length() >= 5 && path.compare(0, 5, "ship/") == 0)
-			return true;
-		if(path.length() >= 9 && path.compare(0, 9, "asteroid/") == 0)
-			return true;
-
-		return false;
-	}
-
 	// Get the character index where the sprite name in the given path ends.
 	size_t NameEnd(const string &path)
 	{
-		// The path always ends in a three-letter extension, ".png" or ".jpg".
-		// In addition, 3 more characters may be taken up by an @2x label or a mask label.
-		bool is2x = Is2x(path);
-		size_t end = path.length() - (is2x ? 7 : 4) - (IsSwizzleMask(path, is2x) ? 3 : 0);
+		auto lastDot = path.find_last_of('.');
+		if(lastDot == string::npos)
+			return 0;
+
+		// Get the name of the file, without the extension and the @2x label.
+		string base = path.substr(0, lastDot);
+		if(base.ends_with("@2x"))
+			base = base.substr(0, base.length() - 3);
+		// In addition, more characters may be taken up by a mask label.
+		if(base.ends_with("@sw"))
+			base = base.substr(0, base.length() - 3);
+
 		// This should never happen, but just in case:
-		if(!end)
+		if(base.empty())
 			return 0;
 
 		// Skip any numbers at the end of the name.
-		size_t pos = end;
+		size_t pos = base.length();
 		while(--pos)
 			if(path[pos] < '0' || path[pos] > '9')
 				break;
 
 		// If there is not a blending mode specifier before the numbers, they
 		// are part of the sprite name, not a frame index.
-		return (IsBlend(path[pos]) ? pos : end);
+		return (IsBlend(path[pos]) ? pos : base.length());
+	}
+
+	// Determine whether the given path is to an @2x image.
+	bool Is2x(const string &path)
+	{
+		return path.substr(NameEnd(path)).find("@2x") != string::npos;
+	}
+
+	// Determine whether the given path is to a swizzle mask.
+	bool IsSwizzleMask(const string &path)
+	{
+		return path.substr(NameEnd(path)).find("@sw") != string::npos;
+	}
```

Comment:
If this stuff is being rewritten, can it not determine the same properties multiple times, multiple different ways?
`NameEnds` checks if the thing is `@2x` and then checks if it is `@sw`. But, we're also calling that function as part of checking if this is `@2x` and `@sw`.

Reply:
I suppose `Is2x` could just not care about the extension, then. I don't really see a case where someone would put `@2x` into a file's name without intending for it to be a 2x sprite.